File: 107. Binary Tree Level Order Traversal II.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def levelOrderBottom(self, root):
        """
        :type root: TreeNode
        :rtype: List[List[int]]
        """
        if not root:
            return []
        queue = [(root, 0)]
        res = [[]]
        while len(queue) > 0:
            temp = queue.pop(0)
            p = temp[0]
            level = temp[1]
            res[len(res) - 1] += [p.val]
            if p.left:
                queue += [(p.left, level + 1)]
            if p.right:
                queue += [(p.right, level + 1)]
            if len(queue) > 0 and level != queue[0][1]:
                res += [[]]
        return res[::-1]

File: 107. Binary Tree Level Order Traversal II.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
#include <queue>
class Solution {
public:
    vector<vector<int>> levelOrderBottom(TreeNode* root) {
        vector<vector<int>> res;
        if (root == NULL)
            return res;
        queue<TreeNode*> nodes;
        queue<int> level;
        res.push_back({});
        nodes.push(root);
        level.push(0);
        while (nodes.size() > 0 and level.size() > 0)
        {
            TreeNode* p = nodes.front();
            nodes.pop();
            int l = level.front();
            level.pop();
            res[res.size()-1].push_back(p->val);
            if (p->left != NULL)
            {
                nodes.push(p->left);
                level.push(l + 1);
            }
            if (p->right != NULL)
            {
                nodes.push(p->right);
                level.push(l + 1);
            }
            if (nodes.size() > 0 && l != level.front())
                res.push_back({});
        }
        reverse(res.begin(), res.end());
        return res;
    }
};
